pkg/kubelet/kubelet.go

```diff
@@ -1944,8 +1948,8 @@ func (kl *Kubelet) deletePod(pod *v1.Pod) error {
 		return fmt.Errorf("deletePod does not allow nil pod")
 	}
 	if !kl.sourcesReady.AllReady() {
-		// If the sources aren't ready, skip deletion, as we may accidentally delete pods
-		// for sources that haven't reported yet.
+		// If the sources aren't ready, skip deletion, the housekeeping will try to remove again while all sources ready
+		kl.waitRemoveQueue.Enqueue(pod)
```

Comment:
Don't we have to remove the pod if an add comes in for that same pod before sources are ready?

Reply:
Make sense, if a pod ADD is not after the same uid pod's REMOVE, it would not trigger the waiting restart.
https://github.com/kubernetes/kubernetes/blob/5f57708c88634c2be2bad3968e1d36ec8c2032c3/pkg/kubelet/pod_workers.go#L615-L617
so if I still call deletePod after it's ADD, it will be not correct, since some one may only want to do a manifest moving out and in to    
let the static pod restart.

Reply:
If we only remove the pod from waiting queue while an add comes in for that some pod, that means we actually did not process the static pod's restart, is it reasonable?  or can we also delay the ADD action to follow the REMOVE while there is a REMOVE in the waiting queue for that some pod?

Reply:
I'm now worried (based on some of your questions) whether we'd just be patching around the issue.  Added a comment below that may suggest a more involved (but ultimately easier to reason about) approach.  Also solicited some feedback from others.